btree/btree.c

```diff
@@ -2921,6 +2925,493 @@ static void ff_capture(struct slot *slot, struct m0_be_tx *tx)
 	M0_BTREE_TX_CAPTURE(tx, seg, h, hsize);
 }
 
+/**
+ *  --------------------------------------------
+ *  Section END - Fixed Format Node Structure
+ *  --------------------------------------------
+ */
+
+/**
+ *  --------------------------------------------
+ *  Section START -
+ *  Variable Sized Key and Values Node Structure
+ *  --------------------------------------------
+ *
+ * Proposed node structure
+ *
+ *
+ *                        +--------------------------+
+ *                        |                          |
+ *                 +------+-----------------------+  |
+ *                 |      |                       |  |
+ *             +---+------+-------------------+   |  |
+ *             |   |      |                   |   |  | DIR                    16   11       3  0
+ * +----------+v-+-v--+---v----+-------------++-+-++-++--+--+--+--------------+----+--------+--+
+ * |          |  |    |        |             |  |  |  |  |  |  |              |    |        |  |
+ * | Node Hdr |K0| K1 |   K3   |             +--+--+--+--+--+--+              | V3 |   V1   |V0|
+ * |          |  |    |        |             |  |  |  |  |  |  |              |    |        |  |
+ * +----------+--+----+--------+-------------++-++-+-++--+--+--+--------------+--^-+-----^--+-^+
+ *            0  5    12                      |  |   |                           |       |    |
+ *                                            +--+---+---------------------------+-------+----+
+ *                                               |   |                           |       |
+ *                                               +---+---------------------------+-------+
+ *                                                   |                           |
+ *                                                   +---------------------------+
+ *
+ *
+ *  The above structure represents the way variable sized keys and values may be
+ *  stored in memory.
+ *  Node Hdr or Node Header will store all the relevant info regarding this node
+ *  type.
+ *  The keys wil be added from the start of the node in an increasing way
+ *  whereas the values will be added from the end of the node in a decreasing 
+ *  way.
+ *  Additionally, we will maintain a directory in the central region of the node
+ *  which will contain the offsets to identify the address of the particular
+ *  key and value.
```

Comment:
Do not forget to mention:
This Directory starts in the center of the empty space between Keys and Values and will float in this region so that if addition of the new record causes either Key or Value to overwrite the Directory then the Directory will need to be shifted to make space for the incoming Key/Value; accordingly the Directory pointer in the Node Header will be updated. If no space exists to add the new Record then an Overflow has happened and the new record cannot be inserted in this node.
The credit calculations for the addition of the Record need to account for the moving of this Directory.

Reply:
Added.